APDS9007.class.nut

```diff
@@ -87,51 +81,74 @@ class APDS9007 {
     }
 
     /**
+     * Get reading
+     * Assumes that the sensor is enabled and enable timeout has passed
+     * @private
+     */
+    function _read() {
+        local Vpin = 0;
+        local Vcc = 0;
+
+        // average several readings for improved precision
+        for (local i = 0; i < _points_per_read; i++) {
+            Vpin += _input_pin.read();
+            Vcc += hardware.voltage();
+        }
+
+        Vpin = (Vpin * 1.0) / _points_per_read;
+        Vcc = (Vcc * 1.0) / _points_per_read;
+        Vpin = (Vpin / 65535.0) * Vcc;
+
+        local Iout = (Vpin / _rload) * 1000000.0; // current in µA
+        return {"brightness" : math.pow(10.0, (Iout / 10.0))};
+    }
+
+    /**
      * Reads and returns a table with a key of brightness
      * containing the ambient light level in Lux.
      *
-     * @param {function(result)|null} cb - callback executed on reading availability
-     * @return {null}
+     * @param {function(result)|null} cb - Callback executed on reading availability. If no callback specified, reading is rejturned
+     * @return {null|{brightness}}
      */
     function read(cb = null) {
-        local result = {};
 
-        if(_enabled) {
-            local Vpin = 0;
-            local Vcc = 0;
+        if (_enabled_at /* sensor is enabled */) {
 
-            // average several readings for improved precision
-            for (local i = 0; i < _points_per_read; i++) {
-                Vpin += _als_pin.read();
-                Vcc += hardware.voltage();
-            }
+            local seconds_since_enabled = (hardware.millis() - _enabled_at) / 1000;
+
+            if (cb /* we're async */) {
 
-            Vpin = (Vpin * 1.0) / _points_per_read;
-            Vcc = (Vcc * 1.0) / _points_per_read;
-            Vpin = (Vpin / 65535.0) * Vcc;
+                if (ENABLE_TIMEOUT < seconds_since_enabled) {
+                    // timeout has passed, we're good to go
+                    imp.wakeup(0, function() { cb(_read()); }.bindenv(this));
```

Comment:
Actually... now storing the reading as you suggested. Because: a delayed one effectively requires double-checking on enable/readiness state since the current execution context could have disabled sensor by the time it clears and callback is executed from the wakeup queue.